alf/networks/disjoint_policy_value_network.py

```diff
@@ -0,0 +1,263 @@
+
+            action_spec (nest of TensorSpec): speficifies the shape and type of
+                the output action. The type of output action distribution is
+                implicitly derived from this.
+
+            encoding_network_ctor (Callable[..., Network]): A constructor that
+                creates the encoding network. Depending whether the encoding
+                network is shared between the value component and the policy
+                component, 1 or 2 encoding network will be created using this
+                constructor.
+
+            is_sharing_encoder (bool): When set to true, the encoding network is
+                shared between the value and the policy component, resulting in
+                a "shared" architecture disjoint network. When set to false, the
+                encoding network is not shared, resulting in a "dual"
+                architecture disjoint network.
+
+            kernel_initializer (Callable): initializer for all the layers
+                excluding the projection net. If none is provided a default
+                xavier_uniform will be used.
+
+            discrete_projection_net_ctor (Callable[..., Network]): constructor
+                that generates a discrete projection network that outputs
+                discrete actions.
+
+            continuous_projection_net_ctor (Callable[..., Network): constructor
+                that generates a continuous projection network that outputs
+                continuous actions.
+
+            name(str): the name of the network
+
+        """
+        super().__init__(input_tensor_spec=observation_spec, name=name)
+
+        # +------------------------------------+
+        # | Step 1: The policy network encoder |
+        # +------------------------------------+
+
+        if kernel_initializer is None:
+            kernel_initializer = torch.nn.init.xavier_uniform_
+
+        self._actor_encoder = encoding_network_ctor(
+            input_tensor_spec=observation_spec,
+            kernel_initializer=kernel_initializer)
+
+        # +------------------------------------------+
+        # | Step 2: Projection for the policy branch |
+        # +------------------------------------------+
+
+        projection_net = _create_projection_net_based_on_action_spec(
+            discrete_projection_net_ctor=discrete_projection_net_ctor,
+            continuous_projection_net_ctor=continuous_projection_net_ctor,
+            input_size=self._actor_encoder.output_spec.shape[0],
+            action_spec=action_spec)
+
+        if alf.nest.is_nested(projection_net):
+            # In the case when the projection net is multi-fold (because actor
+            # spec is a nest), force picking up the parameters inside those
+            # project networks into this DisjointPolicyValueNetwork instance.
+            self._projection_net_module_list = torch.nn.ModuleList(
+                alf.nest.flatten(projection_net))
```

Comment:
No need to have 178-179. `nn.Branch` should be able to handle it correctly.

Reply:
I read the code on `Sequential` and `Branch` and did not find how those in `Branch` can get into the closure. Also did an experiment in the unit test and found that when `action_spec` is a nest, those projection networks' parameters do not propagate back to the `DisjointPolicyValueNetwork`'s state closure.

If this is not intentional, we probably need to fix `Sequential`/`Branch` to add this feature.

Reply:
It's supposed to be handled at line 413 and 160 in networks/containers.py

Reply:
Thanks. I can confirm that after the refactor all the parameters are picked up in `_composition`. Line 413 and 160 in `continers.py` should work.

> Also did an experiment in the unit test and found that when action_spec is a nest, those projection networks' parameters do not propagate back to the DisjointPolicyValueNetwork's state closure.

Could not get back to where the experiment fails, it is likely that I missed the projection parameters when checked with my naked eye.